Improve copy option

fleischie opened the pull request:
Hello. :slightly_smiling_face: 

This PR is a proposal to fix #2: Check, whether there is a `--copy` flag given and copy the required emoji to the clipboard. For this purpose a `choice` parameter can be used to specify the requested index, which defaults to `1`.

There are several checks in place:
- `emoj --copy unicorn` - If `--copy` is given, but no input, try to use `choice` as input.
- `emoj --copy -5 unicorn` - Clamp the selection between 0 and the returned lists `length`.
- `emoj --copy 100 unicorn` - Same as above.

Furthermore without the `--copy` flag but with a `name` parameter `emoji` now utilizes `inquirer` to display an interactive selection for the emoji to copy to the clipboard. I found this is a more usable approach to handle an ambiguous result.


NOTE: While writing this PR message I noticed there are tests in place, that fail. I will update the PR as soon as possible to fix them. You can check out the current solution in the meantime and give notes. :+1:

fleischie commented:
Haha, the failing tests were "only" broken linting rules and I see CI enforced them. So, yeah. All :white_check_mark: :slightly_smiling_face:

SamVerschueren commented on cli.js:
```diff
@@ -9,8 +9,33 @@ const debounce = require('lodash.debounce');
 const mem = require('mem');
 const clipboardy = require('clipboardy');
+const inquirer = require('inquirer');
+
```
Remove empty line

fleischie replied:
I will update the PR when I'm on the computer again. Maybe there's more to review... I can make a review commit then. 😅

fleischie commented:
One of the latest commits removes the whitespace and the other now highlights the selection in color emoji terminals.

<details>I did the initial PR on a linux machine with only emoji-font support, that allowed color-highlighting. Mac on the other hand (which I did the last two commits on) supports color emoji in it's terminal and thus needs another means of highlighting).</details>

sindresorhus commented on cli.js:
```diff
@@ -19,30 +43,78 @@ const debouncer = debounce(cb => cb(), 200);
 
 	Options
-	  --copy -c  Copy the first emoji to the clipboard
+	  --copy -c [choice]  Copy the emoji on index <choice> to the clipboard
```
This doesn't make much sense. You can't know up front which index you'll want.

fleischie replied:
First of all: Sorry for the late reply.
Second: This is a valid concern, although I have to say, that not knowing which index you want also includes the first ony, which is selected by default. This way you can first do `emoj horsie` and then select the one you want with `emoj horsie --copy 3` (for example). I thought this allows users to explicitly select an indexed emojie, if it's not the first one you get via `emoj` on itself.

sindresorhus commented:
I would like to have the same copy interface for both `$ emoj` and `$ emoj input`, so either we need to make this one horizontal, or `emoj` vertical. I think I prefer horizontal.

We could do something like:

```
❯ emoj

› horse
🐴  🐎  🏇  🎠  🚴  😍  🚌
           ^
```

fleischie commented:
Yes, I think having a consistent layout makes sense and is preferable. 🙂 

The current implementation uses `inquirer` and I don't know, whether it has the possibility to inquire horizontally. I will check, though.

sindresorhus commented:
You don't need to use Inquirer. Just capture the arrow keys and move the `^` accordingly. For example: https://github.com/sindresorhus/emoj/blob/2ea3c23667c39c61a5158f2cb815d5490d8b6de4/cli.js#L109

fleischie commented:
Hello. I addressed the changes:

- Remove unnecessary whitespaces. (`cli.js` now passes `xo && ava`.)
- Removed `inquirer` again, but displayed a selection indicator and track the choice.
- I left the copy-index selection in there, because I personally find it a useful feature, and it's backwards compatible, so no short-term harm done. If you ask me to remove it, I will of course, but I did not have the strong feeling of being forced to currently.

As I used the plain response from dango (and not coerce it to a string right away), this might be a breaking change from the known internals.

If you need anything, have any comments/requests/opinions/statements about this PR, just let me know. I am available for discussions/modifications of PR/etc.

sindresorhus commented on cli.js:
```diff
@@ -102,8 +198,11 @@ process.stdin.on('keypress', (ch, key) => {
+			emojiIndex = Math.min(emojiIndex, emojis.length - 1);
+			prevResult = emojis;
+			const indicator = '   '.repeat(emojiIndex);
+			logUpdate(`${pre}${chalk.bold(query.join(''))}\n${stringifyEmojis(prevResult)}\n${indicator}Ʌ`);
```
`Ʌ` => `^`

sindresorhus commented:
When I move between the emojis, the indicator (`^`) disappears for some milliseconds before it appears in the new position. It should not disappear.

fleischie commented:
I addressed the indicator issues, but it seems to be not perfectly aligned in all the terminals I have. Hyper for example spaces the emojis farther away, than 3 characters (at least in my setup). iTerm2 looks good though.

Give me an update on what to address and I'll do it. Thank you for your ongoing patience. 🙇

sindresorhus commented:
Hey, we decided to rewrite this module using [Ink](https://github.com/vadimdemedes/ink) to make the code more readable and maintainable. This, unfortunately, makes your pull request moot. Sorry about that...

fleischie commented:
Yeah, I already saw some days ago in my twitter feed. 😊

Thanks for taking your time, though. I'll be looking forward to the new implementation.
